subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c

```diff
@@ -0,0 +1,928 @@
+			strncpy(saved_job.id, job_id, sizeof(saved_job.id));
+		}
+	}
+
+	ret = settings_save_one(SETTINGS_FULL_FOTA_JOB, &saved_job,
+				sizeof(saved_job));
+	if (ret) {
+		LOG_ERR("settings_save_one failed: %d", ret);
+	}
+
+	return ret;
+}
+
+static void http_fota_handler(const struct fota_download_evt *evt)
+{
+	__ASSERT_NO_MSG(evt != NULL);
+
+	LOG_DBG("%s() - evt %d", log_strdup(__func__), evt->id);
+
+	switch (evt->id) {
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		/* MCUBOOT: download finished, update job status and reboot */
+		current_fota.status = NRF_FOTA_IN_PROGRESS;
+		save_validate_status(current_fota.id, current_fota.type,
+				NRF_FOTA_VALIDATE_PENDING);
+		send_job_update(&current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_PENDING:
+		/* MODEM: update job status and reboot */
+		current_fota.status = NRF_FOTA_IN_PROGRESS;
+		save_validate_status(current_fota.id, current_fota.type,
+				NRF_FOTA_VALIDATE_PENDING);
+		send_job_update(&current_fota);
+		send_event(NRF_FOTA_EVT_ERASE_PENDING, &current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_DONE:
+		/* MODEM: this event is received when the initial
+		 * fragment is downloaded and dfu_target_modem_init() is
+		 * called.
+		 */
+		send_event(NRF_FOTA_EVT_ERASE_DONE, &current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERROR:
+		if (last_fota_dl_evt == FOTA_DOWNLOAD_EVT_ERASE_DONE) {
+			current_fota.status = NRF_FOTA_REJECTED;
+		} else {
+			current_fota.status = NRF_FOTA_FAILED;
+			current_fota.error = NRF_FOTA_ERROR_DOWNLOAD;
+		}
+		save_validate_status(current_fota.id, current_fota.type,
+				NRF_FOTA_VALIDATE_DONE);
+		send_job_update(&current_fota);
+		send_event(NRF_FOTA_EVT_ERROR, &current_fota);
+		cleanup_job(&current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_PROGRESS:
+		current_fota.status = NRF_FOTA_DOWNLOADING;
+		current_fota.dl_progress = evt->progress;
+
+		/* Do not send complete status more than once */
+		if ((current_fota.sent_dl_progress == 100) &&
+		    (current_fota.dl_progress == 100)) {
+			break;
+		}
+
+		/* Reset if new progress is less than previous */
+		if (current_fota.sent_dl_progress >
+		    current_fota.dl_progress) {
+			current_fota.sent_dl_progress = 0;
+		}
+
+		/* Only send progress update when finished and if increment
+		 * threshold is met
+		 */
+		if (current_fota.dl_progress != 100 &&
+		    ((current_fota.dl_progress -
+		      current_fota.sent_dl_progress) <
+		     CONFIG_NRF_CLOUD_FOTA_PROGRESS_PCT_INCREMENT)) {
+			break;
+		}
+
+		current_fota.sent_dl_progress = current_fota.dl_progress;
+		send_job_update(&current_fota);
+		break;
+	default:
+		break;
+	}
+
+	last_fota_dl_evt = evt->id;
+}
+
+static int parse_job(struct nrf_cloud_fota_job * const job)
+{
+	char *save_ptr = NULL;
+	char *end_ptr = &job->mqtt_payload[job->mqtt_payload_size - 1];
+	char *tok;
+
+	/* Job format:
+	 * [“jobExecutionId”,firmwareType,fileSize,”host”,”path”]
+	 *
+	 * Example:
+	 * [“abcd1234”,0,1234,”nrfcloud.com”,"v1/firmwares/appfw.bin"]
```

Comment:
So this looks like a JSON array to me. If cJSON already is used by nRFCloud may I suggest maybe trying to rewrite the parsing logic using cJSON?

Could be that it's not possible it's just a suggestion.

```
int status;
cJSON_ArrayForEach(token, tokens) {
        if (!cJSON_IsNumber(token)) {
            status = 0;
            goto handle_error;
        }
    }

Reply:
probably possible, i can try it out to see how it looks.  i did not consider pulling in a library since the parsing was simple.

Reply:
Yeah, I think cJSON is already in the nRF Cloud library for doing some JSON parsing. So instead of having custom parsing logic (even if simple), it would be nice to abstract it away if possible.